src/main/kotlin/org/taktik/icure/asynclogic/objectstorage/impl/DataAttachmentLoaderImpl.kt

```diff
@@ -0,0 +1,103 @@
+package org.taktik.icure.asynclogic.objectstorage.impl
+
+import java.security.GeneralSecurityException
+import java.security.KeyException
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import org.springframework.core.io.buffer.DataBuffer
+import org.springframework.core.io.buffer.DefaultDataBufferFactory
+import org.springframework.stereotype.Service
+import org.taktik.icure.asyncdao.DocumentDAO
+import org.taktik.icure.asyncdao.GenericDAO
+import org.taktik.icure.asynclogic.objectstorage.DataAttachmentLoader
+import org.taktik.icure.asynclogic.objectstorage.DocumentDataAttachmentLoader
+import org.taktik.icure.asynclogic.objectstorage.DocumentObjectStorage
+import org.taktik.icure.asynclogic.objectstorage.DocumentObjectStorageMigration
+import org.taktik.icure.asynclogic.objectstorage.IcureObjectStorage
+import org.taktik.icure.asynclogic.objectstorage.IcureObjectStorageMigration
+import org.taktik.icure.asynclogic.objectstorage.contentBytesOfNullable
+import org.taktik.icure.entities.Document
+import org.taktik.icure.entities.base.HasDataAttachments
+import org.taktik.icure.entities.embed.DataAttachment
+import org.taktik.icure.properties.ObjectStorageProperties
+import org.taktik.icure.security.CryptoUtils
+import org.taktik.icure.security.CryptoUtils.isValidAesKey
+import org.taktik.icure.security.CryptoUtils.keyFromHexString
+
+class DataAttachmentLoaderImpl<T : HasDataAttachments<T>>(
+	private val dao: GenericDAO<T>,
+	private val icureObjectStorage: IcureObjectStorage<T>,
+	private val icureObjectStorageMigration: IcureObjectStorageMigration<T>,
+	private val objectStorageProperties: ObjectStorageProperties
+): DataAttachmentLoader<T> {
+	override fun contentFlowOf(
+		target: T,
+		retrieveAttachment: T.() -> DataAttachment
+	): Flow<DataBuffer> = target.retrieveAttachment().let { attachment ->
+		attachment.contentFlowFromCacheOrLoad { doLoadFlow(target, attachment) }
+	}
+
+	override suspend fun contentBytesOf(
+		target: T,
+		retrieveAttachment: T.() -> DataAttachment
```

Comment:
I'm not sure if with this you mean target has no `DataAttachment` or target has `DataAttachment`s where the content is not available yet, so I will address both these situations.


`retrieveAttachment` will be invoked on `target`, and since the `retrieveAttachment` function is implemented by the method user it is his job to ensure that it returns a non-null attachment. For example if the user knows for sure that `target` has an attachment with key `”preview"` he could invoke the method as follows:
```kotlin
contentBytesOf(target) { dataAttachments.getValue("preview") }
```
however, if target does not actually have a data attachment with key "preview" the `retrieveAttachment` function will throw a `NoSuchElementException`, which won't be handled by the `contentBytesOf` method. 
If the user is not sure whether a certain `DataAttachment` exists or not, because for example he wants to use the same `retrieveAttachment` function on multiple `target`s it would be better to use the nullable version of the method:
```kotlin
contentBytesOfNullable(target: T?, retrieveAttachment: T.() -> DataAttachment?): ByteArray?
```
In this case `retrieveAttachment` may return null if there is no `DataAttachment`, then the result will also be null.


In the cases where the content is not available, for example because it was not fully uploaded to object storage and it was created by another icure user, so we don't have the attachment stored locally, the appropriate exception will be thrown.